### ISSUE TITLE
Path normalization controls

### ISSUE THREAD

**Issue by htuch** [envoyproxy/envoy#6589]
The fix for CVE-2019-9900, provided a coarse grained ability to opt-in/out of path normalization. That is, you could either normalize, in which case both normalization was used for matching and transforming the path to the upstream, or not normalize.

Ideally we provide finer grained controls similar to Nginx, where users can opt to normalize for match independent of transforming the path to the upstream.

CC @PiotrSikora 

Action item for CVE-2019-9900

**Comment by htuch** [comment]
Bumping priority, since today we still have normalization disabled by default.

**Comment by htuch** [comment]
Once we have this, we should ensure that normalization occurs by default.

**Comment by twghu** [comment]
Looking at this along with https://github.com/envoyproxy/envoy/issues/6588

**Comment by twghu** [comment]
Work on finding a library for normalisation progessed on https://github.com/envoyproxy/envoy/issues/6588 seems tied to the solution for this.  Will monitor and action pending outcome of same.

**Comment by htuch** [comment]
@twghu related but not dependent, I think we can add in the controls with the existing `PathUtil` interface.

**Comment by jonnysgomes** [comment]
Hi @htuch , I'm using Istio and I'm having an issue related with path traversal. The Envoy version is 1.12. Istio brings the normalize path configuration enabled by default, as we can see [here](https://github.com/istio/istio/pull/14191/files). Even with that configuration enabled, I still can navigate through the url path, for instance:

`https://my.domain.net/myservice/../otherplace`

According with [this](https://github.com/envoyproxy/envoy/commit/7ed6d2187df94c4cb96f7dccb8643bf764af2ccb) link, Envoy already has fixed that. I've checked out the `normalize_path` flag and it's enabled. By my understanding, the url above should be resolved as `https://my.domain.net/myservice/`. So, it's could be a bad Envoy behavior or am I missing somethig?

**Comment by PiotrSikora** [comment]
@jonnysgomes normalized `https://my.domain.net/myservice/../otherplace` is `https://my.domain.net/otherplace`, not `https://my.domain.net/myservice`.

Without normalization, `/myservice/../otherplace` is matched against `/myservice` prefix, which is potentially a security issue, since you're really accessing `/otherplace`.

With normalization, `/myservice/../otherplace` is matched against `/otherplace` prefix, which is correct.

**Comment by twghu** [comment]
Implementation:

Requirements:
1. All existing behavior will be preserved.  
2. Normalization should occur by default ( following previous comment htuch 22/1 ).
3. Code changes will be restricted to the upstream connection only with no changes to routing code to eliminate introduction of new security/function bugs.

Changes:
1.  Add a configuration option to HttpConnectionManager _normalize_path_for_upstream_ with default to value of _normalize_path_ (support existing behavior). 
2. Add a new header entry OriginalPathForUpstream, This will default to Path() as per current functionality.  If _normalize_path_for_upstream_ is false only then will OriginalPathForUpstream be passed to upstream non-normalized.    This allows Envoy the option to use a normalized path from downstream to perform route matching while optionally allowing the non-normalized path to be passed through to upstream.  The header entry is not forwarded to upstream and is private to Envoy.  The addition of this new entry provides visibility to other internal functionality that may have an interest in the upstream path while minimizing code changes in the code base.
3.  If the _normalize_path_ configuration is false, then no normalization occurs for upstream (support existing behavior) by default.  If _normalize_path_for_upstream_ is set to true then the upstream path can be normalized independently.

**Comment by htuch** [comment]
This seems like a good design to me. @PiotrSikora @mattklein123 thoughts? 

You'll need to be careful about dumping `OriginalPathForUpstream` on the wire by accident, or reaching upstream code with it not set (e.g. from an sync client).

**Comment by mattklein123** [comment]
Sorry I'm having a hard time wrapping my head around ^ without seeing it in context with the existing config. Can you potentially do a config/docs only PR and we can discuss the proposal there?

**Comment by twghu** [comment]
Hey Matt See https://github.com/envoyproxy/envoy/pull/11111

**Comment by howardjohn** [comment]
Is this only about path? Or would something like https://github.com/envoyproxy/envoy/pull/14491 fall in scope? Essentially allowing the original Host header (with port) to be forwarded while routing without the port internally. The current behavior forwards the modified host header, much like the path normalization

**Comment by howardjohn** [comment]
Another similar question as above: Is header case normalization in scope? I can open a new issue if its both desired and out of scope. https://github.com/istio/istio/issues/30579 for reference

**Comment by htuch** [comment]
I think #8463 is tracking this one @howardjohn.

**Comment by asraa** [comment]
For observability, if we split this path we may want to default log the original path (that contains the most information, normalized paths can be recovered from the original). We should add formatters like `%UPSTREAM_PATH%` or `%NORMALIZED_PATH%` in case users want a way to view the upstream or internal normalization for matching.

**Comment by htuch** [comment]
Yes, I think original path should be default as anyone doing audit logging will want to see that.

**Comment by ramaraochavali** [comment]
@PiotrSikora @htuch It seems the normalize_path does not completely percent decode the path. For example the path `/jetty-status%2f../jetty-status` normalizes to `/jetty-status%2f../jetty-status`  where as we expecting it to normalize to `/jetty-status` - Is this expected?  This path `/jetty-status/../jetty-status` normalizes correctly to `/jetty-status`. If this is as expected, what do we need to support that?
